### PR TITLE
updated build.sh file

### PR DESCRIPTION
**Pull request by theNerdFromSiliconValley**
1) User-friendly
2) Fixed code signing
3) Better DMG creation, User can save 24.99$

### CODE REVIEW

**Comment by theNerdFromSiliconValley** [comment]
Johan Boss, I have added text instructions before user Team Details according to your advice, thanks for that

**Comment by johankool** [comment]
I propose to keep the DropDMG as an option as I still think that is the option that SURFnet should use to create a DMG for distribution. It for example also signs the dmg for distribution, shows the version in the Finder window etc. The script should not install create-dmg until that option is chosen. 

Though the entire need for a DMG for users using this build script to build the app for themselves still eludes me. It would be entirely reasonable for the user to just say no to creating the DMG and take the build app and distributed it by whichever method they deem fit for their purpose.

Could you also put back the pretty colors? :grin:

Have you actually confirmed that the app works when entering a team ID other than SURFnet or Egeniq?

**Comment by theNerdFromSiliconValley** [comment]
Hi Johan Boss,
{Have you actually confirmed that the app works when entering a team ID other than SURFnet or Egeniq?}
Here is the DMG file, when entering a team ID other than SURFnet or Egeniq. 
Link: https://drive.google.com/open?id=1zyDAs1S41W9eX3kr2uuKZ9XD-wAtm4Xw
you can change the background.png file to change the design. 

DropDMG's background text and the color is just an image. If you the check Top startup's DMG file like Dropbox, they all use their designers to do that. 

Best
Robin

**Comment by johankool** [comment]
Hi Robin,

Thanks for confirming it works with other Team IDs.

I understand how the background works in DMGs. Nevertheless I want the option to use DropDMG to remain as an option in the script. 

Thanks,

Johan (not a boss, just Johan btw)

**Comment by fkooman** [comment]
> I propose to keep the DropDMG as an option as I still think that is the option that SURFnet should use to create a DMG for distribution. It for example also signs the dmg for distribution, shows the version in the Finder window etc. The script should not install create-dmg until that option is chosen.

So DropDMG is more secure than create-dmg? And create-dmg does not have this functionality? Does this actually protect the user in any way if they just click next (or drag the icon to "Applications") in either case?

> Though the entire need for a DMG for users using this build script to build the app for themselves still eludes me. It would be entirely reasonable for the user to just say no to creating the DMG and take the build app and distributed it by whichever method they deem fit for their purpose.

The whole point of this is that any other party can create the user-installable software in the exact same way as eduVPN/LC does that. Reducing the number of (additional) proprietary components as much as possible seems like a good idea in that case!

**Comment by theNerdFromSiliconValley** [comment]
We are signing the DMG after creating it using OS's default tool
codesign -f -s "$SIGNINGIDENTITY" $INSTALLERFILENAME

File: openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh [hunk 1]
```diff
@@ -1523,23 +1496,20 @@ readonly TB_RESOURCES_PATH="${ARG_TB_PATH}/Contents/Resources"
 #
 #     LEASEWATCHER_PLIST_PATH and REMOVE_LEASEWATCHER_PLIST are passed to the other scripts via the scutil State:/Network/OpenVPN mechanism
 
-# NOT! The functionality from the above comment is disabled
```
Comment (johankool) [inline openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh]: The changes here from line 1526 to 1543 should be reverted for the lease watch functionality to keep working.

File: openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh [hunk 2]
```diff
@@ -1503,7 +1476,7 @@ readonly CONFIG_PATH_DASHES_SLASHES="$( echo "${TBCONFIG}" | sed -e 's/-/--/g' |
 readonly SCRIPT_LOG_FILE="/Library/Application Support/eduVPN/Logs/${CONFIG_PATH_DASHES_SLASHES}.script.log"
 
 readonly TB_RESOURCES_PATH="${ARG_TB_PATH}/Contents/Resources"
-# readonly FREE_PUBLIC_DNS_SERVERS_LIST_PATH="${TB_RESOURCES_PATH}/FreePublicDnsServersList.txt"
+readonly FREE_PUBLIC_DNS_SERVERS_LIST_PATH="${TB_RESOURCES_PATH}/FreePublicDnsServersList.txt"
```
Comment (johankool) [inline openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh]: This resources is not in the app. This change should be reverted.

File: openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh [hunk 3]
```diff
@@ -1304,33 +1280,33 @@ logDnsInfo() {
...
+				done
+				if ${knownDnsServerNotFound} ; then
+					logMessage "NOTE: The DNS servers do not include any free public DNS servers known to eduVPN. This may cause DNS queries to fail or be intercepted or falsified even if they are directed through the VPN. Specify only known public DNS servers or DNS servers located on the VPN network to avoid such problems."
+				else
+					if ${unknownDnsServerFound} ; then
+						logMessage "NOTE: The DNS servers include one or more free public DNS servers known to eduVPN and one or more DNS servers not known to eduVPN. If used, the DNS servers not known to eduVPN may cause DNS queries to fail or be intercepted or falsified even if they are directed through the VPN. Specify only known public DNS servers or DNS servers located on the VPN network to avoid such problems."
+					else
+						logMessage "The DNS servers include only free public DNS servers known to eduVPN."
+					fi
+				fi
```
Comment (johankool) [inline openvpn-2.4.6-openssl-1.1.0h/client.up.eduvpn.sh]: This block of code about knownPublicDnsServers was purposely disabled.

**Comment by theNerdFromSiliconValley** [comment]
Since the pull request was pending for build.sh, my new commit automatically added to last pull request. Sorry about that.  I am making another pull request for up and down script

File: build.sh [hunk 4]
```diff
@@ -16,12 +37,30 @@ echo ""
 echo "Which signing identity do you want to use?"
 echo "1. SURFnet B.V. (ZYJ4TZX4UU)"
 echo "2. Egeniq (E85CT7ZDJC)"
-echo "3. Other"
-read -p "0-9?" choice
+echo "3. Enter own Team(This isn't just a random name, but it must be exact as in their signing identity): "
```
Comment (johankool) [inline build.sh]: `3. Enter own Team(This isn't just a random name, but it must be exact as in their signing identity): "` --> `3. Other "`

File: build.sh [hunk 5]
```diff
@@ -16,12 +37,30 @@ echo ""
...
+read -p "1-3?" choice
+
+
+# Enter custom Team ID and Team Name.
+if [ "$choice" == 3  ]
+then
```
Comment (johankool) [inline build.sh]: Add line: `echo "Your Team ID and Team Name must match exactly with the signing identity in your keychain."`

File: build.sh [hunk 6]
```diff
@@ -49,7 +88,7 @@ esac
 FILENAME="$TARGET-$VERSION"
 
 echo ""
-echo "$(tput setaf 2)Bootstrapping dependencies using carthage$(tput sgr 0)"
```
Comment (johankool) [inline build.sh]: Please put back color

File: build.sh [hunk 7]
```diff
@@ -58,42 +97,44 @@ EOF
 XCODE_XCCONFIG_FILE="${PWD}/Carthage/64bit.xcconfig" carthage bootstrap --cache-builds --platform macOS
 
 echo ""
-echo "$(tput setaf 2)Building and archiving$(tput sgr 0)"
```
Comment (johankool) [inline build.sh]: Please put back color

File: build.sh [hunk 8]
```diff
@@ -58,42 +97,44 @@ EOF
...
+echo "Building and archiving"
 xcodebuild archive -project eduVPN.xcodeproj -scheme $TARGET -archivePath $FILENAME.xcarchive DEVELOPMENT_TEAM=$TEAMID
 
 echo ""
-echo "$(tput setaf 2)Exporting$(tput sgr 0)"
```
Comment (johankool) [inline build.sh]: Please put back color

File: build.sh [hunk 9]
```diff
@@ -58,42 +97,44 @@ EOF
...
+echo "Exporting"
 /usr/libexec/PlistBuddy -c "Set :teamID \"$TEAMID\"" ExportOptions.plist
 xcodebuild -exportArchive -archivePath $FILENAME.xcarchive -exportPath $FILENAME -exportOptionsPlist ExportOptions.plist
 
 echo ""
-echo "$(tput setaf 2)Re-signing up and down scripts$(tput sgr 0)"
```
Comment (johankool) [inline build.sh]: Please put back color

File: build.sh [hunk 10]
```diff
@@ -58,42 +97,44 @@ EOF
...
 echo ""
-read -p "Create disk image (requires DropDMG license) (y/n)?" choice
-case "$choice" in
-  y|Y ) ;;
-  n|N ) exit 0;;
-  * ) echo "Invalid response."; exit 0;;
-esac
 
-echo ""
-echo "$(tput setaf 2)Creating a disk image$(tput sgr 0)"
```
Comment (johankool) [inline build.sh]: Tell user what you are doing

File: build.sh [hunk 11]
```diff
@@ -58,42 +97,44 @@ EOF
...
-# The configuration eduVPN can be used for all products
-echo "Using: dropdmg --config-name \"eduVPN\" --signing-identity=\"$SIGNINGIDENTITY\" \"$FILENAME/$PRODUCT\""
-dropdmg --config-name "eduVPN" --signing-identity="$SIGNINGIDENTITY" "$FILENAME/$PRODUCT"
+INSTALLERFILENAME="eduVPN-Installer$(date +"%Y-%m-%d-%H:%M:%S.")dmg"
```
Comment (johankool) [inline build.sh]: Don't hardcode eduVPN

File: build.sh [hunk 12]
```diff
@@ -58,42 +97,44 @@ EOF
...
-echo ""
-echo "$(tput setaf 2)Creating app cast XML$(tput sgr 0)"
-DISTRIBUTIONPATH="../eduvpn-macos-distrib"
-# Assumptions are being made about the location of this script
-# Also, this often fails due to extended attribute
-echo "Using: $DISTRIBUTIONPATH/generate_appcast $DISTRIBUTIONPATH/dsa_priv.pem $DISTRIBUTIONPATH/updates/"
-$DISTRIBUTIONPATH/generate_appcast $DISTRIBUTIONPATH/dsa_priv.pem $DISTRIBUTIONPATH/updates/
+
+create-dmg \
+--volname "eduVPN" \
```
Comment (johankool) [inline build.sh]: Don't hard code eduVPN. Call it Let's Connect if applicable

File: build.sh [hunk 13]
```diff
@@ -58,42 +97,44 @@ EOF
...
+--volicon "icon.icns" \
+--background "background.png" \
+--window-pos 490 350 \
+--window-size 490 350 \
+--icon-size 100 \
+--icon "eduVPN.app" 100 155 \
```
Comment (johankool) [inline build.sh]: Don't hard code app name

File: build.sh [hunk 14]
```diff
@@ -58,42 +97,44 @@ EOF
...
-echo "$(tput setaf 2)Creating app cast XML$(tput sgr 0)"
-DISTRIBUTIONPATH="../eduvpn-macos-distrib"
-# Assumptions are being made about the location of this script
-# Also, this often fails due to extended attribute
-echo "Using: $DISTRIBUTIONPATH/generate_appcast $DISTRIBUTIONPATH/dsa_priv.pem $DISTRIBUTIONPATH/updates/"
-$DISTRIBUTIONPATH/generate_appcast $DISTRIBUTIONPATH/dsa_priv.pem $DISTRIBUTIONPATH/updates/
+
+create-dmg \
+--volname "eduVPN" \
+--volicon "icon.icns" \
```
Comment (johankool) [inline build.sh]: Don't use eduVPN logo for the volume.

File: build.sh [hunk 15]
```diff
@@ -58,42 +97,44 @@ EOF
...
+--background "background.png" \
+--window-pos 490 350 \
+--window-size 490 350 \
+--icon-size 100 \
+--icon "eduVPN.app" 100 155 \
+--hide-extension "eduVPN.app" \
+--app-drop-link 370 155 \
+ $INSTALLERFILENAME \
+$FILENAME"/eduVPN.app"
```
Comment (johankool) [inline build.sh]: Don't hard code app name

**Comment by johankool** [comment]
I'd still say this change still isn't okay. It does not create a DMG similar to what it was before. The background doesn't look like what it was before. I think the look before was better.

**Comment by fkooman** [comment]
I'd say with the build script we should focus on functionality, reproducibility, free software tools and a way for everyone to easily create a working eduVPN/LC installer, even if it doesn't look exactly the same instead of bikeshedding the background color of the installer.

**Comment by fkooman** [comment]
I mean, unless there are clear technical/security reasons for not switching of course!

**Comment by johankool** [comment]
My main issue is that the build script now hard code bits it shouldn't.

File: build.sh (hunk 4 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: Hi Johan, So you want to change the text here to other?

File: build.sh (hunk 5 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: Done

File: build.sh (hunk 6 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: That is I want to leave to the designer. I think you want white. But me and you are just coder, so designer can better decide on this

**Comment by theNerdFromSiliconValley** [comment]
I will try to remove as much as hard code possible

**Comment by theNerdFromSiliconValley** [comment]
Johan, from the begining of the script we have hardcoded everything,


`echo "Which target do you want to build?"
echo "1. eduVPN"
echo "2. Let's Connect!"
read -p "1-2?" choice
case "$choice" in
  1 ) TARGET="eduVPN"; PRODUCT="eduVPN.app";;
  2 ) TARGET="LetsConnect"; PRODUCT="Let's Connect!.app";;
  * ) echo "Invalid response."; exit 0;;
esac

echo ""
echo "Which signing identity do you want to use?"
echo "1. SURFnet B.V. (ZYJ4TZX4UU)"
echo "2. Egeniq (E85CT7ZDJC)"
echo "3. Other "
read -p "1-3?" choice`

I can put some variable in the beginning so that people can change this all easily.

File: build.sh (hunk 14 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: What is the problem with putting eduVPN icon here?

**Comment by johankool** [comment]
The bit you quote is not hard coded, but the ones you use for creating the DMG are, like the volume name etc.

File: build.sh (hunk 14 above)
Comment (johankool) [reply build.sh]: The icon in icon.icns is for an app, not for a volume. Also, it's the wrong icon if you build for Let's Connect target.

**Comment by theNerdFromSiliconValley** [comment]
Hi Johan, can you please review the hardcode part?

File: build.sh [hunk 16]
```diff
@@ -1,12 +1,19 @@
 #!/bin/bash
-echo "Build Script for eduVPN (and derivatives)"
+
+APPNAME = "eduVPN"
```
Comment (johankool) [inline build.sh]: No, this is not what I meant. This doesn't help at all. You should use the $TARGET when creating the DMG.

Comment (theNerdFromSiliconValley) [reply build.sh]: case "$choice" in
  1 ) TARGET="$APPNAME"; PRODUCT="$APPNAME.app";;
  2 ) TARGET="LetsConnect"; PRODUCT="Let's Connect!.app";;
  * ) echo "Invalid response."; exit 0;;

Comment (theNerdFromSiliconValley) [reply build.sh]: we have put that in TARGET too :)

Comment (johankool) [reply build.sh]: Yes, but that doesn't solve anything. The $APPNAME variable doesn't help. What is needed is that where eduVPN is used as a string in the creation of the DMG, the $TARGET variable is used.

Comment (theNerdFromSiliconValley) [reply build.sh]: Ok Johan , while creating DMG, the TARGET variable in used

**Comment by theNerdFromSiliconValley** [comment]
Hi Johan, if it ok now, can you please approve the PR?

File: build.sh (hunk 6 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: I have put back white color

File: build.sh (hunk 14 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: Done, since script can now put two different icon

File: build.sh (hunk 7 above)
Comment (theNerdFromSiliconValley) [reply build.sh]: white color is back

**Comment by johankool** [comment]
See PR #182 instead.

**Pull request closed without merging**